### ISSUE TITLE
File::file needs equality operator

### ISSUE THREAD

**Issue by martukas** [ess-dmsc/h5cpp#29]
Should be done using Object_id

**Comment by eugenwintersberger** [comment]
I will have a look on this

**Comment by eugenwintersberger** [comment]
Currently the test fails as it produces two different id though there is only one file and a symbolic link to it. I would have expected that this should pass. Have I misunderstood something?

**Comment by eugenwintersberger** [comment]
@martukas can you please have a look on this?

**Comment by eugenwintersberger** [comment]
After having a look on the actual implementation we maybe could resolve this problem by modifying the ObjectId class. 
1. replace file_name member by two boost::filesystem::path instances
    * `boost::filesystem::path access_path_`
    * and `boost::filesystem::path hard_link_path_`
2. an HDF5 file remembers the path which was used to access the file. When creating an ObjectID we 
    store this informatoin in `access_path_` and check whether or not this is a link. If it is a symbolic link 
    we try to resolve this an store this result in `hard_link_path_`. If it is no symbolic link we change set
    `hard_link_path_` to the same value as `access_path_`.
3. for the `==` operator we use now the `hard_link_path_` to check for the equality of the files. 

There is still an issue if two hard links refer to the same file. It is possible to do this however I have never seem someone actually doing it.

**Comment by martukas** [comment]
Resurrected the ObjectId tests, which were outdated and stashed away in main.cpp. Now they are back in our test package. A thorough review of ObjectId tests revealed some underlying issues. Everything is now better commented, and one badly designed test has been corrected. The general summary is as such:

object_address is stable across all scenarios
file_number:
- not equal if same file closed and reopened
- identifies the file that owns object, not owner of link (in case of ext link)
file_name:
- not equal if file opened via symbolic link (h5 allows opening same file twice)
- not equal if object opened via external link vs. original node

Conclusions:
file_number adequately reflects identity, unless file has been closed and reopened. file_name does not adequately reflect identity. Resolving symbolic link would solve part of the problem, but would also need to somehow dereference external links to identify the name of the object-owning file. Therefore, I have changed the implementation of ObjectID to use file_number rather than file_name, which is at this point more practical. If we want to correctly identify the owning file of a node for the purpose of automatically choosing the correct type of object link to use, perhaps this can be a different ticket and will come up anyways in unit testing for linking.

The test now passes.
Once the implementation for OnbjectID was changed, it should have passed, but was still failing. The issue was the old symbolic link not being deleted. Creating a new h5 with the truncate flag would correctly replace  the old one, but boost's create_symlink would through an exception if it already existed.
